edb/server/protocol/auth_ext/http.py

```diff
@@ -1039,32 +1156,118 @@ async def handle_reset_password(
     ) -> None:
         data = self._get_data_from_request(request)
 
-        _check_keyset(data, {"provider", "reset_token", "password"})
-        reset_token = data['reset_token']
-        password = data['password']
-        email_password_client = email_password.Client(db=self.db)
+        reset_token = data.get('reset_token')
+        email = data.get('email')
+        code = data.get('code')
+        password = data.get('password')
+        provider = data.get('provider')
+
+        if not password:
+            raise errors.InvalidData('Missing "password" in reset request')
+
+        if not provider:
+            raise errors.InvalidData('Missing "provider" in reset request')
 
         allowed_redirect_to = self._maybe_make_allowed_url(
             data.get("redirect_to")
         )
 
+        email_password_client = email_password.Client(db=self.db)
+
         try:
-            token = jwt.ResetToken.verify(
-                reset_token,
-                self.signing_key,
-            )
+            if reset_token:
+                # Token-based password reset (existing flow)
```

Comment:
These "existing flow" and "new flow" comments are actually quite useful for me here but maybe not so much as long-term comments.

Might be worth splitting into functions here also?